File: R/point.R
#' Vector of points
#'
#' Points are the fundamental unit in geometry, from which other geometric
#' primitives can be constructed. Points support less than and greater than
#' operations which are handled lexicographically. This means that point vectors
#' can also be sorted and ranked. Subtracking a point from a point gives a
#' vector. Adding and subtracting a vector to a point translates the point by
#' the vector.
#'
#' @param ... Various input. See the Constructor section.
#' @param default_dim The dimensionality when constructing an empty vector
#' @param x A point vector or an object to convert to it
#'
#' @return a `euclid_point` vector
#'
#' @section Constructors:
#' **2 dimensional points**
#' - Providing two numeric vector will construct points with those x and y
#'   coordinate.
#' - Providing a vector will construct a point at the location the vector
#'   points to.
#' - Providing a weighted point vector will create a points at the same
#'   locations without weights.
#'
#' **3 dimensional points**
#' - Providing three numeric vector will construct points with those x, y, and z
#'   coordinate.
#' - Providing a vector will construct a point at the location the vector
#'   points to.
#' - Providing a weighted point vector will create a points at the same
#'   locations without weights.
#'
#' @export
#'
#' @examples
#' num1 <- exact_numeric(runif(5))
#' num2 <- exact_numeric(runif(5))
#' num3 <- exact_numeric(runif(5))
#'
#' # 2 dimensions
#' p <- point(num1, num2)
#' p
#'
#' plot(p)
#'
#' # 3 dimensions
#' point(num1, num2, num3)
#'
#' # Standard R vectors are automatically converted to exact_numeric vectors
#' point(runif(5), runif(5))
#'
#' # Convert points to vectors
#' as_vec(p)
#'
#' # Arithmetic
#' # Translate by adding vector
#' p + vec(3, 7)
#'
#' # Create vector by subtracting points
#' p[1:2] - p[3:4]
#'
#' # Sorting etc.
#' sort(p)
#' min(p)
#' cummax(p)
#' rank(p)
#' p[1:2] < p[3:4]
#'
point <- function(..., default_dim = 2) {
  inputs <- validate_constructor_input(...)

  if (length(inputs) == 0) {
    return(new_point_empty(default_dim))
  }

  vectors <- inputs[vapply(inputs, is_vec, logical(1))]
  numbers <- inputs[vapply(inputs, is_exact_numeric, logical(1))]
  w_points <- inputs[vapply(inputs, is_weighted_point, logical(1))]

  if (length(vectors) == 1) {
    new_point_from_vec(vectors[[1]])
  } else if (length(w_points) == 1) {
    new_point_from_wp(w_points[[1]])
  } else if (length(numbers) == 2) {
    new_point_from_xy(numbers[[1]], numbers[[2]])
  } else if (length(numbers) == 3) {
    new_point_from_xyz(numbers[[1]], numbers[[2]], numbers[[3]])
  } else {
    rlang::abort("Don't know how to construct points from the given input")
  }
}
#' @rdname point
#' @export
is_point <- function(x) inherits(x, "euclid_point")


# Conversion --------------------------------------------------------------

#' @rdname point
#' @export
as_point <- function(x) {
  UseMethod("as_point")
}
#' @export
as_point.default <- function(x) {
  rlang::abort("Don't know how to convert the input to points")
}
#' @export
as_point.euclid_point <- function(x) x

#' @export
as_vec.euclid_point <- function(x) {
  vec(x)
}

# Operators ---------------------------------------------------------------

#' @export
geometry_op_plus.euclid_point <- function(e1, e2) {
  if (!is_vec(e2)) {
    rlang::abort("`+` is only defined for vectors")
  }
  if (dim(e1) != dim(e2)) {
    rlang::abort("`e1` and `e2` must have the same number of dimensions")
  }
  if (length(e1) == 0 || length(e2) == 0) {
    return(new_point_empty(dim(e1)))
  }
  if (dim(e1) == 2) {
    restore_euclid_vector(point_2_add_vector(get_ptr(e1), get_ptr(e2)), e1)
  } else {
    restore_euclid_vector(point_3_add_vector(get_ptr(e1), get_ptr(e2)), e1)
  }
}
#' @export
geometry_op_minus.euclid_point <- function(e1, e2) {
  if (!is_vec(e2) && !is_point(e2)) {
    rlang::abort("`-` is only defined for vectors and points")
  }
  if (dim(e1) != dim(e2)) {
    rlang::abort("`e1` and `e2` must have the same number of dimensions")
  }
  if (length(e1) == 0 || length(e2) == 0) {
    if (is_point(e2)) {
      return(new_vector_empty(dim(e1)))
    } else {
      return(new_point_empty(dim(e1)))
    }
  }
  if (dim(e1) == 2) {
    if (is_point(e2)) {
      new_geometry_vector(point_2_sub_point(get_ptr(e1), get_ptr(e2)))
    } else {
      restore_euclid_vector(point_2_sub_vector(get_ptr(e1), get_ptr(e2)), e1)
    }
  } else {
    if (is_point(e2)) {
      new_geometry_vector(point_3_sub_point(get_ptr(e1), get_ptr(e2)))
    } else {
      restore_euclid_vector(point_3_sub_vector(get_ptr(e1), get_ptr(e2)), e1)
    }
  }
}
#' @export
geometry_op_less.euclid_point <- function(e1, e2) {
  if (!is_point(e2)) {
    rlang::abort("`<` is only defined for points")
  }
  if (dim(e1) != dim(e2)) {
    rlang::abort("`e1` and `e2` must have the same number of dimensions")
  }
  if (length(e1) == 0 || length(e2) == 0) {
    return(logical(0))
  }
  if (dim(e1) == 2) {
    point_2_less(get_ptr(e1), get_ptr(e2))
  } else {
    point_3_less(get_ptr(e1), get_ptr(e2))
  }
}
#' @export
geometry_op_greater.euclid_point <- function(e1, e2) {
  if (!is_point(e2)) {
    rlang::abort("`>` is only defined for points")
  }
  if (dim(e1) != dim(e2)) {
    rlang::abort("`e1` and `e2` must have the same number of dimensions")
  }
  if (length(e1) == 0 || length(e2) == 0) {
    return(logical(0))
  }
  if (dim(e1) == 2) {
    point_2_greater(get_ptr(e1), get_ptr(e2))
  } else {
    point_3_greater(get_ptr(e1), get_ptr(e2))
  }
}

# Summaries ---------------------------------------------------------------

#' @export
geometry_summary_min.euclid_point <- function(x, na_rm) {
  if (dim(x) == 2) {
    point_2_min(get_ptr(x), na_rm)
  } else {
    point_3_min(get_ptr(x), na_rm)
  }
}
#' @export
geometry_summary_max.euclid_point <- function(x, na_rm) {
  if (dim(x) == 2) {
    point_2_max(get_ptr(x), na_rm)
  } else {
    point_3_max(get_ptr(x), na_rm)
  }
}

# Math --------------------------------------------------------------------

#' @export
geometry_math_cummin.euclid_point <- function(x) {
  if (dim(x) == 2) {
    point_2_cummin(get_ptr(x))
  } else {
    point_3_cummin(get_ptr(x))
  }
}
#' @export
geometry_math_cummax.euclid_point <- function(x) {
  if (dim(x) == 2) {
    point_2_cummax(get_ptr(x))
  } else {
    point_3_cummax(get_ptr(x))
  }
}

# Miscellaneous vector support --------------------------------------------

#' @export
sort.euclid_point <- function(x, decreasing = FALSE, na.last = NA, ...) {
  if (dim(x) == 2) {
    restore_euclid_vector(point_2_sort(get_ptr(x), decreasing, na.last), x)
  } else {
    restore_euclid_vector(point_3_sort(get_ptr(x), decreasing, na.last), x)
  }
}
#' @export
xtfrm.euclid_point <- function(x) {
  if (dim(x) == 2) {
    point_2_rank(get_ptr(x))
  } else {
    point_3_rank(get_ptr(x))
  }
}
#' @export
range.euclid_point <- function(..., na.rm = FALSE) {
  input <- do.call(c, list(...))
  c(min(input, na.rm = na.rm), max(input, na.rm = na.rm))
}

# Internal constructors ---------------------------------------------------

new_point_empty <- function(dim) {
  if (dim == 2) {
    new_geometry_vector(create_point_2_empty())
  } else {
    new_geometry_vector(create_point_3_empty())
  }
}
new_point_from_xy <- function(x, y) {
  new_geometry_vector(create_point_2_x_y(get_ptr(x), get_ptr(y)))
}
new_point_from_xyz <- function(x, y, z) {
  new_geometry_vector(create_point_3_x_y_z(get_ptr(x), get_ptr(y), get_ptr(z)))
}
new_point_from_vec <- function(p) {
  if (dim(p) == 2) {
    new_geometry_vector(create_point_2_vec(get_ptr(p)))
  } else {
    new_geometry_vector(create_point_3_vec(get_ptr(p)))
  }
}
new_point_from_wp <- function(p) {
  if (dim(p) == 2) {
    new_geometry_vector(create_point_2_wp(get_ptr(p)))
  } else {
    new_geometry_vector(create_point_3_wp(get_ptr(p)))
  }
}

File: src/point.cpp
#include "point.h"
#include "point_w.h"
#include "exact_numeric.h"
#include "vector.h"

#include <cpp11/external_pointer.hpp>
#include <cpp11/logicals.hpp>
#include <cpp11/integers.hpp>

[[cpp11::register]]
point2_p create_point_2_empty() {
  std::vector<Point_2> vec;
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
point2_p create_point_2_x_y(exact_numeric_p x, exact_numeric_p y) {
  std::vector<Point_2> vec;
  vec.reserve(x->size());
  for (size_t i = 0; i < x->size(); ++i) {
    if (!(*x)[i] || !(*y)[i]) {
      vec.push_back(Point_2::NA_value());
      continue;
    }
    vec.emplace_back((*x)[i], (*y)[i]);
  }
  point2 *result(new point2(vec));

  return {result};
}

[[cpp11::register]]
point2_p create_point_2_vec(vector2_p p) {
  std::vector<Point_2> vec;
  vec.reserve(p->size());
  for (size_t i = 0; i < p->size(); ++i) {
    if (!(*p)[i]) {
      vec.push_back(Point_2::NA_value());
      continue;
    }
    vec.emplace_back((*p)[i].x(), (*p)[i].y());
  }
  point2 *result(new point2(vec));

  return {result};
}

[[cpp11::register]]
point2_p create_point_2_wp(point_w2_p p) {
  std::vector<Point_2> vec;
  vec.reserve(p->size());
  for (size_t i = 0; i < p->size(); ++i) {
    if (!(*p)[i]) {
      vec.push_back(Point_2::NA_value());
      continue;
    }
    vec.push_back((*p)[i].point());
  }
  point2 *result(new point2(vec));

  return {result};
}

[[cpp11::register]]
point2_p point_2_add_vector(point2_p x, vector2_p y) {
  std::vector<Point_2> vec = (*x) + y->get_storage();
  point2 *result(new point2(vec));

  return {result};
}

[[cpp11::register]]
point2_p point_2_sub_vector(point2_p x, vector2_p y) {
  std::vector<Point_2> vec = (*x) - y->get_storage();
  point2 *result(new point2(vec));

  return {result};
}

[[cpp11::register]]
vector2_p point_2_sub_point(point2_p x, point2_p y) {
  std::vector<Vector_2> vec = (*x) - y->get_storage();
  vector2 *result(new vector2(vec));

  return {result};
}

[[cpp11::register]]
cpp11::writable::logicals point_2_less(point2_p x, point2_p y) {
  return (*x) < y->get_storage();
}

[[cpp11::register]]
cpp11::writable::logicals point_2_greater(point2_p x, point2_p y) {
  return (*x) > y->get_storage();
}

[[cpp11::register]]
point2_p point_2_sort(point2_p x, bool decreasing, cpp11::logicals na_last) {
  std::vector<Point_2> vec = x->sort(decreasing, na_last);
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
cpp11::writable::integers point_2_rank(point2_p x) {
  return x->rank();
}

[[cpp11::register]]
point2_p point_2_min(point2_p x, bool na_rm) {
  std::vector<Point_2> vec = x->min(na_rm);
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
point2_p point_2_max(point2_p x, bool na_rm) {
  std::vector<Point_2> vec = x->max(na_rm);
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
point2_p point_2_cummin(point2_p x) {
  std::vector<Point_2> vec = x->cummin();
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
point2_p point_2_cummax(point2_p x) {
  std::vector<Point_2> vec = x->cummax();
  point2 *result(new point2(vec));
  return {result};
}

[[cpp11::register]]
point3_p create_point_3_empty() {
  std::vector<Point_3> vec;
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
point3_p create_point_3_x_y_z(exact_numeric_p x, exact_numeric_p y, exact_numeric_p z) {
  std::vector<Point_3> vec;
  vec.reserve(x->size());
  for (size_t i = 0; i < x->size(); ++i) {
    if (!(*x)[i] || !(*y)[i] || !(*z)[i]) {
      vec.push_back(Point_3::NA_value());
      continue;
    }
    vec.emplace_back((*x)[i], (*y)[i], (*z)[i]);
  }
  point3 *result(new point3(vec));

  return {result};
}

[[cpp11::register]]
point3_p create_point_3_vec(vector3_p p) {
  std::vector<Point_3> vec;
  vec.reserve(p->size());
  for (size_t i = 0; i < p->size(); ++i) {
    if (!(*p)[i]) {
      vec.push_back(Point_3::NA_value());
      continue;
    }
    vec.emplace_back((*p)[i].x(), (*p)[i].y(), (*p)[i].z());
  }
  point3 *result(new point3(vec));

  return {result};
}

[[cpp11::register]]
point3_p create_point_3_wp(point_w3_p p) {
  std::vector<Point_3> vec;
  vec.reserve(p->size());
  for (size_t i = 0; i < p->size(); ++i) {
    if (!(*p)[i]) {
      vec.push_back(Point_3::NA_value());
      continue;
    }
    vec.push_back((*p)[i].point());
  }
  point3 *result(new point3(vec));

  return {result};
}

[[cpp11::register]]
point3_p point_3_add_vector(point3_p x, vector3_p y) {
  std::vector<Point_3> vec = (*x) + y->get_storage();
  point3 *result(new point3(vec));

  return {result};
}

[[cpp11::register]]
point3_p point_3_sub_vector(point3_p x, vector3_p y) {
  std::vector<Point_3> vec = (*x) - y->get_storage();
  point3 *result(new point3(vec));

  return {result};
}

[[cpp11::register]]
vector3_p point_3_sub_point(point3_p x, point3_p y) {
  std::vector<Vector_3> vec = (*x) - y->get_storage();
  vector3 *result(new vector3(vec));

  return {result};
}

[[cpp11::register]]
cpp11::writable::logicals point_3_less(point3_p x, point3_p y) {
  return (*x) < y->get_storage();
}

[[cpp11::register]]
cpp11::writable::logicals point_3_greater(point3_p x, point3_p y) {
  return (*x) > y->get_storage();
}

[[cpp11::register]]
point3_p point_3_sort(point3_p x, bool decreasing, cpp11::logicals na_last) {
  std::vector<Point_3> vec = x->sort(decreasing, na_last);
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
cpp11::writable::integers point_3_rank(point3_p x) {
  return x->rank();
}

[[cpp11::register]]
point3_p point_3_min(point3_p x, bool na_rm) {
  std::vector<Point_3> vec = x->min(na_rm);
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
point3_p point_3_max(point3_p x, bool na_rm) {
  std::vector<Point_3> vec = x->max(na_rm);
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
point3_p point_3_cummin(point3_p x) {
  std::vector<Point_3> vec = x->cummin();
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
point3_p point_3_cummax(point3_p x) {
  std::vector<Point_3> vec = x->cummax();
  point3 *result(new point3(vec));
  return {result};
}

[[cpp11::register]]
cpp11::writable::logicals point_collinear(geometry_vector_base_p x, geometry_vector_base_p y, geometry_vector_base_p z) {
  size_t output_size = std::max(std::max(x->size(), y->size()), z->size());
  cpp11::writable::logicals result;
  result.reserve(output_size);
  if (x->dimensions() == 2) {
    auto p1 = get_vector_of_geo<Point_2>(*x);
    auto p2 = get_vector_of_geo<Point_2>(*y);
    auto p3 = get_vector_of_geo<Point_2>(*z);
    for (size_t i = 0; i < output_size; ++i) {
      if (!p1[i % p1.size()] || !p2[i % p2.size()] || !p3[i % p3.size()]) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::collinear(p1[i % p1.size()], p2[i % p2.size()], p3[i % p3.size()]));
    }
  } else {
    auto p1 = get_vector_of_geo<Point_3>(*x);
    auto p2 = get_vector_of_geo<Point_3>(*y);
    auto p3 = get_vector_of_geo<Point_3>(*z);
    for (size_t i = 0; i < output_size; ++i) {
      if (!p1[i % p1.size()] || !p2[i % p2.size()] || !p3[i % p3.size()]) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::collinear(p1[i % p1.size()], p2[i % p2.size()], p3[i % p3.size()]));
    }
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::logicals point_coplanar(point3_p x, point3_p y, point3_p z, point3_p t) {
  size_t output_size = std::max(std::max(std::max(x->size(), y->size()), z->size()), t->size());
  cpp11::writable::logicals result;
  result.reserve(output_size);
  for (size_t i = 0; i < output_size; ++i) {
    if (!(*x)[i % x->size()] || !(*y)[i % y->size()] || !(*z)[i % z->size()] || !(*t)[i % t->size()]) {
      result.push_back(NA_LOGICAL);
      continue;
    }
    result.push_back((Rboolean) CGAL::coplanar((*x)[i % x->size()], (*y)[i % y->size()], (*z)[i % z->size()], (*t)[i % t->size()]));
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::logicals point_ordered(geometry_vector_base_p x, geometry_vector_base_p y, geometry_vector_base_p z) {
  size_t output_size = std::max(std::max(x->size(), y->size()), z->size());
  cpp11::writable::logicals result;
  result.reserve(output_size);
  if (x->dimensions() == 2) {
    auto p1 = get_vector_of_geo<Point_2>(*x);
    auto p2 = get_vector_of_geo<Point_2>(*y);
    auto p3 = get_vector_of_geo<Point_2>(*z);
    for (size_t i = 0; i < output_size; ++i) {
      if (!p1[i % p1.size()] || !p2[i % p2.size()] || !p3[i % p3.size()]) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::are_ordered_along_line(p1[i % p1.size()], p2[i % p2.size()], p3[i % p3.size()]));
    }
  } else {
    auto p1 = get_vector_of_geo<Point_3>(*x);
    auto p2 = get_vector_of_geo<Point_3>(*y);
    auto p3 = get_vector_of_geo<Point_3>(*z);
    for (size_t i = 0; i < output_size; ++i) {
      if (!p1[i % p1.size()] || !p2[i % p2.size()] || !p3[i % p3.size()]) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::are_ordered_along_line(p1[i % p1.size()], p2[i % p2.size()], p3[i % p3.size()]));
    }
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::logicals point_ordered_along(geometry_vector_base_p x) {
  cpp11::writable::logicals result;
  result.reserve(x->size() - 2);
  if (x->dimensions() == 2) {
    auto vec = get_vector_of_geo<Point_2>(*x);
    for (size_t i = 0; i < x->size() - 2; ++i) {
      Point_2 p1 = vec[i];
      Point_2 p2 = vec[i + 1];
      Point_2 p3 = vec[i + 2];
      if (!p1 || !p2 || !p3) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::are_ordered_along_line(p1, p2, p3));
    }
  } else {
    auto vec = get_vector_of_geo<Point_3>(*x);
    for (size_t i = 0; i < x->size() - 2; ++i) {
      Point_3 p1 = vec[i];
      Point_3 p2 = vec[i + 1];
      Point_3 p3 = vec[i + 2];
      if (!p1 || !p2 || !p3) {
        result.push_back(NA_LOGICAL);
        continue;
      }
      result.push_back((Rboolean) CGAL::are_ordered_along_line(p1, p2, p3));
    }
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::logicals point_turns_left(point2_p x, point2_p y, point2_p z) {
  size_t output_size = std::max(std::max(x->size(), y->size()), z->size());
  cpp11::writable::logicals result;
  result.reserve(output_size);
  for (size_t i = 0; i < output_size; ++i) {
    if (!(*x)[i % x->size()] || !(*y)[i % y->size()] || !(*z)[i % z->size()]) {
      result.push_back(NA_LOGICAL);
      continue;
    }
    result.push_back((Rboolean) CGAL::left_turn((*x)[i % x->size()], (*y)[i % y->size()], (*z)[i % z->size()]));
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::logicals point_turns_right(point2_p x, point2_p y, point2_p z) {
  size_t output_size = std::max(std::max(x->size(), y->size()), z->size());
  cpp11::writable::logicals result;
  result.reserve(output_size);
  for (size_t i = 0; i < output_size; ++i) {
    if (!(*x)[i % x->size()] || !(*y)[i % y->size()] || !(*z)[i % z->size()]) {
      result.push_back(NA_LOGICAL);
      continue;
    }
    result.push_back((Rboolean) CGAL::right_turn((*x)[i % x->size()], (*y)[i % y->size()], (*z)[i % z->size()]));
  }
  return result;
}

[[cpp11::register]]
cpp11::writable::integers point_turns(point2_p x) {
  cpp11::writable::integers result;
  result.reserve(x->size() - 2);
  for (size_t i = 0; i < x->size() - 2; ++i) {
    Point_2 p1 = (*x)[i];
    Point_2 p2 = (*x)[i + 1];
    Point_2 p3 = (*x)[i + 2];
    if (!p1 || !p2 || !p3) {
      result.push_back(R_NaInt);
      continue;
    }
    if (CGAL::left_turn(p1, p2, p3)) {
      result.push_back(1);
      continue;
    }
    if (CGAL::right_turn(p1, p2, p3)) {
      result.push_back(-1);
      continue;
    }
    result.push_back(0);
  }
  return result;
}
